Web Inspector: Replace deprecated String.prototype.substr with substring and at for single parameters

stwrt opened the pull request:
#### e2e6e3b758f6e8966f8a3d05b5b6234eea60dea5
<pre>
Web Inspector: Replace deprecated String.prototype.substr with substring and at for single parameters
<a href="https://bugs.webkit.org/show_bug.cgi?id=303832">https://bugs.webkit.org/show_bug.cgi?id=303832</a>
<a href="https://rdar.apple.com/problem/166140736">rdar://problem/166140736</a>

Reviewed by Devin Rousso.

String.prototype.substr is deprecated (Annex B) and should be replaced
with modern alternatives.

Using at in places with negative values, and using substring in places we can prove will never be below 0.

Only handling the single parameter case, because the two parameter function calls have different syntax:
String.prototype.substr ( start, length )
String.prototype.slice ( start, end )

* Source/WebInspectorUI/UserInterface/Base/MIMETypeUtilities.js:
(WI.fileExtensionForFilename):
* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(get return):
* Source/WebInspectorUI/UserInterface/Controllers/HARBuilder.js:
(WI.HARBuilder.port):
* Source/WebInspectorUI/UserInterface/Models/Cookie.js:
(WI.Cookie.parseSetCookieResponseHeader):
* Source/WebInspectorUI/UserInterface/Models/Gradient.js:
(WI.Gradient.stopsWithComponents):
* Source/WebInspectorUI/UserInterface/Test/TestHarness.js:
(TestHarness.sanitizeURL):
(TestHarness.sanitizeStackFrame):

Canonical link: <a href="https://commits.webkit.org/304181@main">https://commits.webkit.org/304181@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22df26b71fb56aeef2773b4109e6fb98729a5063

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7250 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46061 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142326 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86714 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d1138bb0-3433-4fab-9581-2c82f805c823) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7860 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7096 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/97eb2317-879b-4f24-9668-cbea7f9581ee) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137756 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5549 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120836 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83849 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d10a1463-a7b5-4d5d-851c-9959f6376167) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5383 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2996 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2916 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38987 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145021 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6925 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39563 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6989 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5810 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111733 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28344 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5222 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117115 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60858 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6970 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35294 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6761 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70552 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6997 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6870 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Andrew-Fryer commented on Source/WebInspectorUI/UserInterface/Controllers/HARBuilder.js:
```diff
@@ -310,7 +310,7 @@ WI.HARBuilder = class HARBuilder
         if (!index)
```
This check is falsy for `-1`.
I wonder if this should really be returning `undefined` when `":"` is not in the string...?

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Utilities.js:
```diff
@@ -857,7 +857,7 @@ Object.defineProperty(String.prototype, "truncateStart",
         if (this.length <= maxLength)
             return this;
-        return ellipsis + this.substr(this.length - maxLength + 1);
+        return ellipsis + this.slice(this.length - maxLength + 1);
```
i think most of these actually cant be negative due to prior checks, so id kinda prefer we use `substring` instead since i normally associate `slice` with arrays

stwrt replied:
You are right. Looking through most of the changes they cannot be negative.

There is one spot that we can prove uses a negative that we would need to keep to match the original behavior.

```
if (component.length && component[0].slice(-1) === "%")
```

I'll update the other spots to substring.

stwrt committed:
Web Inspector: Replace deprecated String.prototype.substr with substring and at for single parameters
https://bugs.webkit.org/show_bug.cgi?id=303832
rdar://problem/166140736

Reviewed by Devin Rousso.

String.prototype.substr is deprecated (Annex B) and should be replaced
with modern alternatives.

Using at in places with negative values, and using substring in places we can prove will never be below 0.

Only handling the single parameter case, because the two parameter function calls have different syntax:
String.prototype.substr ( start, length )
String.prototype.slice ( start, end )

* Source/WebInspectorUI/UserInterface/Base/MIMETypeUtilities.js:
(WI.fileExtensionForFilename):
* Source/WebInspectorUI/UserInterface/Base/Utilities.js:
(get return):
* Source/WebInspectorUI/UserInterface/Controllers/HARBuilder.js:
(WI.HARBuilder.port):
* Source/WebInspectorUI/UserInterface/Models/Cookie.js:
(WI.Cookie.parseSetCookieResponseHeader):
* Source/WebInspectorUI/UserInterface/Models/Gradient.js:
(WI.Gradient.stopsWithComponents):
* Source/WebInspectorUI/UserInterface/Test/TestHarness.js:
(TestHarness.sanitizeURL):
(TestHarness.sanitizeStackFrame):

Canonical link: https://commits.webkit.org/304181@main
